File: sorting-algorithms/bubble-sort/c-sharp/bubble_sort_all_versions.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{
    class SortingAlgorithms
    {
        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            // Execute all three versions of the bubble sort algorithms
            Console.WriteLine("\n### Bubble sort version 1 (for loops) ###");
            BubbleSortVersion1(GetTestData());
            
            Console.WriteLine("\n### Bubble sort version 2 (while and for loops) ###");
            BubbleSortVersion2(GetTestData());
            
            Console.WriteLine("\n### Bubble sort version 3 (while and for loops improved) ###");
            BubbleSortVersion3(GetTestData());
        }


        // Returns the test data for the algorithm
        public static int[] GetTestData()
        {
            //int[] testItems = new int[] {80, 64, 50, 43, 35, 21, 7, 3, 2}; // Least sorted
            //int[] testItems = new int[] {2, 3, 7, 35, 43, 21, 50, 64, 80}; // Nearly sorted
            //int[] testItems = new int[] {2, 3, 7, 21, 35, 43, 50, 64, 80}; // Sorted
            int[] testItems = new int[] {43, 21, 2, 50, 3, 80, 35, 7, 64}; // Random
            
            Console.Write($"Data: ");
            Console.WriteLine("[{0}]", string.Join(", ", testItems));

            return testItems;
        }


        // An inefficient bubble sort that uses nested for loops
        public static void BubbleSortVersion1(int[] items)
        {
            int numItems = items.Length;
            int passNum; // Declared outside of the for loop for testing below

            for (passNum = 1; passNum < numItems; passNum++) {
                int index = 0; // Declared outside of the for loop for testing below
                for (index = 0; index < numItems - 1; index++) {
                    if (items[index] > items[index + 1]) {
                        int temp = items[index];
                        items[index] = items[index + 1];
                        items[index + 1] = temp;
                    }
                }
                // Testing
                Console.Write($"Pass {passNum}: ");
                Console.Write("[{0}]", string.Join(", ", items)); 
                Console.WriteLine($"  Comparisons: {index+1}");
            }
        }


        // A quite efficient bubble sort that stops if the items are sorted
        public static void BubbleSortVersion2(int[] items)
        {
            int numItems = items.Length;
            bool swapped = true;
            int passNum = 1; // Testing

            while (swapped == true) {
                swapped = false;
                int index = 0; // Declared outside of the for loop for testing below
                for (index = 0; index < numItems - 1; index++) {
                    if (items[index] > items[index + 1]) {
                        int temp = items[index];
                        items[index] = items[index + 1];
                        items[index + 1] = temp;
                        swapped = true;
                    }
                }
                // Testing
                Console.Write($"Pass {passNum}: ");
                Console.Write("[{0}]", string.Join(", ", items)); 
                Console.WriteLine($"  Comparisons: {index+1}");
                
                passNum = passNum + 1;
            }
        }


        // A more efficient bubble sort that reduces the number of comparisons per pass
        public static void BubbleSortVersion3(int[] items)
        {
            int numItems = items.Length;
            bool swapped = true;
            int passNum = 1;

            while (swapped == true) {
                swapped = false;
                int index = 0; // Declared outside of the for loop for testing below
                for (index = 0; index < numItems - passNum; index++) {
                    if (items[index] > items[index + 1]) {
                        int temp = items[index];
                        items[index] = items[index + 1];
                        items[index + 1] = temp;
                        swapped = true;
                    }
                }
                // Testing
                Console.Write($"Pass {passNum}: ");
                Console.Write("[{0}]", string.Join(", ", items)); 
                Console.WriteLine($"  Comparisons: {index+1}");
                
                passNum = passNum + 1;
            }
        }

        
    }
}

File: sorting-algorithms/bubble-sort/python/bubble_sort_all_versions.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


def bubble_sort_version_1(items):
    """An inefficient bubble sort that uses nested for loops"""
    num_items = len(items)
    
    for pass_num in range(1, num_items): # pseudocode: num_items - 1
        for index in range(0, num_items - 1): # pseudocode: num_items - 2
            if items[index] > items[index + 1]:
                temp = items[index]
                items[index] = items[index + 1]
                items[index + 1] = temp
        print(f"Pass {pass_num}: {items}   Comparisons: {index+1}") # Testing


def bubble_sort_version_2(items):
    """A quite efficient bubble sort that stops if the items are sorted"""
    num_items = len(items)
    swapped = True
    pass_num = 1 # Testing
    
    while swapped:
        swapped = False
        for index in range(0, num_items - 1): # pseudocode: num_items - 2
            if items[index] > items[index + 1]:
                temp = items[index]
                items[index] = items[index + 1]
                items[index + 1] = temp
                swapped = True
        print(f"Pass {pass_num}: {items}   Comparisons: {index+1}") # Testing
        pass_num = pass_num + 1 # Testing


def bubble_sort_version_3(items):
    """A more efficient bubble sort that reduces the number of comparisons per pass"""
    num_items = len(items)
    swapped = True
    pass_num = 1
    
    while swapped:
        swapped = False # Assume sorted
        for index in range(0, num_items - pass_num): # pseudocode: num_items - 1 - pass_num
            if items[index] > items[index + 1]:
                temp = items[index]
                items[index] = items[index + 1]
                items[index + 1] = temp
                swapped = True
        print(f"Pass {pass_num}: {items}   Comparisons: {index+1}") # Testing
        pass_num = pass_num + 1


def get_test_data():
    """Returns the test data for the algorithm"""
    #test_items = [80,64,50,43,35,21,7,3,2] # Least sorted
    #test_items = [2,3,7,35,43,21,50,64,80] # Nearly sorted
    #test_items = [2,3,7,21,35,43,50,64,80] # Sorted
    test_items = [43,21,2,50,3,80,35,7,64] # Random

    print(f"Data: {test_items}")
    
    return test_items


def main():
    """Execute all three versions of the bubble sort algorithms"""
    print("\n### Bubble sort version 1 (for loops) ###")
    bubble_sort_version_1(get_test_data())
    
    print("\n### Bubble sort version 2 (while and for loops) ###")
    bubble_sort_version_2(get_test_data())
    
    print("\n### Bubble sort version 3 (while and for loops improved) ###")
    bubble_sort_version_3(get_test_data())


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()
